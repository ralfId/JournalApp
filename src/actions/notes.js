import Swal from "sweetalert2";


import { db } from "../firebase/firebaseConfig";
import { loadNotes } from "../helpers/loadNotes";
import { uploadImage } from "../helpers/uploadImage";
import { types } from "../types/types";

export const startNewNote = () => {

    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const docRef = await db.collection(`${uid}/journal/notes`).add(newNote)
                .catch(e => {
                    Swal.fire('Saved', e.message, 'success');
                });
            dispatch(activeNote(docRef.id, newNote));
            dispatch(addNewNote(docRef.id, newNote));
        } catch (error) {
            console.log('startNewNote>>>>>>>>>>>>>>>>>' + error)
        }
    }
}

export const activeNote = (id, note) => ({
    type: types.notesSetActiveNote,
    payload: { id, ...note }
})

export const addNewNote = (id, note) => ({
    type: types.notesAddNewNote,
    payload: { id, ...note }
})


export const startLoadingNotes = (id) => {
    return async (dispatch) => {
        try {
            const notes = await loadNotes(id);
            dispatch(setNotes(notes));
        } catch (error) {
            console.log('startLoadingNotes >>>>>>' + error)
        }
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoadNotes,
    payload: notes
})

export const startSavingNote = (note) => {

    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        if (!note.imageUrl) {
            delete note.imageUrl
        }

        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFireStore)
            .catch(e => {
                console.log(e.message)
                Swal.fire('Saved', e.message, 'success');
            });
        dispatch(refreshUpdatedNote(note.id, noteToFireStore));


        Swal.fire({
            position: 'top-end',
            icon: 'success',
            title: 'Your note has been saved',
            text: `${note.title}`,
            showConfirmButton: false,
            timer: 1500
        })
    }
}

export const refreshUpdatedNote = (id, note) => ({
    type: types.notesRefrshUpdatedNote,
    payload: { id: id, note: { id, ...note } }
})

export const startUploadingImage = (file) => {
    return async (dispatch, getState) => {

        try {
            const activeNote = getState().notes.active;

            Swal.fire({
                title: 'Loading...',
                text: 'Please wait...',
                allowOutsideClick: false,
                showConfirmButton: false,
                didOpen: () => { Swal.showLoading() }
            })
            
            const imageFileUrl = await uploadImage(file);
            activeNote.imageUrl = imageFileUrl;

            dispatch(startSavingNote(activeNote));

            Swal.close();
        } catch (error) {
            console.log('startUploadingImage >>>>>>>>>>>>' + error)
        }

    }
}

export const startDeleteNote = (id) => {
    return async (dispath, getState) => {

        const uid = getState().auth.uid;

        Swal.fire({
            title: 'Loading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            showConfirmButton: false,
            didOpen: () => { Swal.showLoading() }
        })
        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispath(deleteNote(id));

        Swal.close()

    }
}

export const deleteNote = (id) => ({
    type: types.notesDeleteNote,
    payload: { id }
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaningNotes
})
