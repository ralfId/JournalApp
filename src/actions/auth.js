import { firebase, googleAuthProvider } from '../firebase/firebaseConfig'
import Swal from 'sweetalert2'

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { notesLogout } from './notes';


export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(startLoading());

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {

                dispatch(login(user.uid, user.displayName));

                dispatch(finishLoading());

            })
            .catch(e => {
                
                dispatch(finishLoading());

                Swal.fire({
                    icon: 'error',
                    title: '',
                    text: e.message,
                  })
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
        
            .then(async ({ user }) => {

                await user.updateProfile({ displayName: name })
                dispatch(login(user.uid, user.displayName))

            }).catch(error =>{
                 Swal.fire('Error', error.message, 'error')
                })
    }
}


export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => { dispatch(login(user.uid, user.displayName)) })
    }
}


export const login = (uid, displayName) => (
    {
        type: types.login,
        payload: { uid, displayName }
    })

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();
        dispatch(logout())
        dispatch(notesLogout());
    }
}
export const logout = (params) => ({ type: types.logout})

