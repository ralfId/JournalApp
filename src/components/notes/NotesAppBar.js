import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSavingNote, startUploadingImage } from '../../actions/notes';

export const NotesAppBar = () => {


    const dispatch = useDispatch();

    const { active } = useSelector(state => state.notes)

    const handleSaveNote = () => {
        // console.log(active)
        dispatch(startSavingNote(active));
    }

    const handleSelectImage = () => {
        document.querySelector('#fileSelector').click()
    }
    
    const handleOpenFileSelector = (e) => {
        const file  = e.target.files[0]
        dispatch(startUploadingImage(file));
    }
    

    return (
        <div className="notes__appbar ">
            <span> 28 de agosto de 2021</span>
            <input id='fileSelector' type="file" onChange={handleOpenFileSelector} style={{display: 'none'}}/>
            <div>
                <button className="btn" onClick={handleSelectImage}>
                    Picture
                </button>
                <button className="btn" onClick={handleSaveNote}>
                    Save
                </button>
            </div>
        </div>
    )
}
