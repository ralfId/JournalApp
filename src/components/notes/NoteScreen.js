import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleteNote } from '../../actions/notes'
import { useForm } from '../../hooks/useForms'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { active: note } = useSelector(state => state.notes)
    const [formValues, handleInputChange, reset] = useForm(note);
    const { id, title, body } = formValues;

    const activeId = useRef(note.id);

    useEffect(() => {

        if (note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch(activeNote(formValues.id, { ...formValues }))
    }, [formValues, dispatch]);

    const handleDeleteNote = () => {
        dispatch(startDeleteNote(id));
    }
    
    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">

                <label className="mt-5 mb-1" htmlFor="title">Title</label>
                <input type="text"
                    placeholder="Title" name="title"
                    className="notes__title-input" value={title} onChange={handleInputChange} />

                <label className="mt-5 mb-1" htmlFor="body">Description</label>
                <textarea className="notes__textarea " name="body"
                    placeholder="What happened today?"
                    value={body} onChange={handleInputChange}></textarea>

                <div className="notes__image">
                    {
                        (note.imageUrl)
                        &&
                        (
                            <img src={note.imageUrl} alt="fantastic" />
                        )
                    }
                </div>
            </div>

            <button className="btn btn-danger" onClick={handleDeleteNote}>
                 Delete
            </button>

        </div>

    )
}
