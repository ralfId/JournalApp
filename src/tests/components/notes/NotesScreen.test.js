import React from 'react';
import { mount } from "enzyme";
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk'

import '@testing-library/jest-dom';
import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';


jest.mock('../../../actions/notes', () => (({
    activeNote: jest.fn()
})));

const middlewares = [thunk]
const mockStore = configureStore(middlewares)


const initState = {
    auth: {
        uid: '1',
        displayName: 'Rafael'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 1,
            title: 'asd',
            body: '123'
        }
    }

};

let store = mockStore(initState)
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <NoteScreen />
    </Provider>
)


describe('pruebas en NotesScreen.js', () => {
    
    test('debe hacer match con el snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    test('debe de llamarse activeNote', () => {
        
        wrapper.find('input[name="title"]').simulate('change',{
            target:{
                name: 'title',
                value:'new title'
            }
        })

        expect(activeNote).toHaveBeenCalled();
        expect(activeNote).toHaveBeenCalledWith(
            1,{
                body:'123',
                id:1,
                title:'new title'
            }
        );
    })
    
});