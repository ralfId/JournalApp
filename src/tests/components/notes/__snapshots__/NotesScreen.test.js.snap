// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en NotesScreen.js debe hacer match con el snapshot 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar "
  >
    <span>
       28 de agosto de 2021
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <label
      className="mt-5 mb-1"
      htmlFor="title"
    >
      Title
    </label>
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      type="text"
      value="asd"
    />
    <label
      className="mt-5 mb-1"
      htmlFor="body"
    >
      Description
    </label>
    <textarea
      className="notes__textarea "
      name="body"
      onChange={[Function]}
      placeholder="What happened today?"
      value="123"
    />
    <div
      className="notes__image"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
